homeassistant/components/flashforge/sensor.py

```diff
@@ -0,0 +1,179 @@
+from __future__ import annotations
+
+from ffpp.Printer import Printer, ToolHandler
+
+from homeassistant.components.sensor import STATE_CLASS_MEASUREMENT, SensorEntity
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import DEVICE_CLASS_TEMPERATURE, PERCENTAGE, TEMP_CELSIUS
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+
+from .const import DEFAULT_NAME, DOMAIN
+from .data_update_coordinator import FlashForgeDataUpdateCoordinator
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the available FlashForge sensors platform."""
+
+    # Get coordinator and unique id.
+    coordinator: FlashForgeDataUpdateCoordinator = hass.data[DOMAIN][
+        config_entry.entry_id
+    ]["coordinator"]
+    printer: Printer = hass.data[DOMAIN][config_entry.entry_id]["printer"]
+    device_id = config_entry.unique_id or ""
+
+    entities: list[SensorEntity] = []
+
+    # Create temperature sensors.
+    if printer.connected:
+        types = ["now", "target"]
+        for tool in printer.extruder_tools:
+            for temp_type in types:
+                entities.append(
+                    FlashForgeTemperatureSensor(
+                        coordinator,
+                        tool.name,
+                        temp_type,
+                        printer.extruder_tools,
+                        device_id,
+                    )
+                )
+        for tool in printer.bed_tools:
+            for temp_type in types:
+                entities.append(
+                    FlashForgeTemperatureSensor(
+                        coordinator,
+                        tool.name,
+                        temp_type,
+                        printer.bed_tools,
+                        device_id,
+                    )
+                )
+
+    entities.append(FlashForgeStatusSensor(coordinator, device_id))
+    entities.append(FlashForgeJobPercentageSensor(coordinator, device_id))
+
+    async_add_entities(entities)
+
+
+class FlashForgeSensorBase(CoordinatorEntity, SensorEntity):
+    """Representation of an FlashForge sensor."""
+
+    coordinator: FlashForgeDataUpdateCoordinator
+
+    def __init__(
+        self,
+        coordinator: FlashForgeDataUpdateCoordinator,
+        sensor_type: str,
+        device_id: str,
+    ) -> None:
+        """Initialize a new FlashForge sensor."""
+        super().__init__(coordinator)
+        self._device_id = device_id
+        self._attr_name = f"{DEFAULT_NAME} {sensor_type}"
+        self._attr_unique_id = f"{sensor_type}-{device_id}"
+
+    @property
+    def device_info(self):
+        """Device info."""
+        return self.coordinator.device_info
+
+
+class FlashForgeStatusSensor(FlashForgeSensorBase):
```

Comment:
Instead of making classes for each sensor to handle the native value and available methods, use `SensorEntityDescription`s. You can add a lambda to those that would handle that.

It would greatly simplify and reduce the amount of code needed. You could, for example, look at the `tailscale` integration (which does this).

Reply:
I guess that i was looking in some older integration. Thanks for pointing me in right direction.